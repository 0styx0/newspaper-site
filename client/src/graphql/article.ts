import gql from 'graphql-tag';

const ArticleQuery = gql`
    query ArticleQuery($issue: Int!, $url: String!) {
        articles(issue: $issue, url: $url) {
            id,
            article,
            canEdit,
            tags,
            comments {
                id
                content,
                dateCreated,
                canDelete,
                author {
                    fullName,
                    profileLink
                }
            }
        }
    }

`;

const ArticleCreate = gql`
    mutation ArticleCreate($tags: [String], $url: String!, $article: String!) {
        createArticle(tags: $tags, url: $url, article: $article) {
            issue
            url
        }
    }
`;

export {
    ArticleQuery,
    ArticleCreate
};
