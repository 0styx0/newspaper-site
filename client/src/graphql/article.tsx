import { gql } from 'react-apollo';

const ArticleQuery = gql`
    query ArticleQuery($issue: Int!, $url: String!) {
        articles(issue: $issue, url: $url) {
            id,
            article,
            canEdit,
            comments {
                id
                content,
                dateCreated,
                canEdit,
                author {
                    fullName,
                    profileLink
                }
            }
            tags {
                all
            }
        }
    }

`;

export {
    ArticleQuery
};
