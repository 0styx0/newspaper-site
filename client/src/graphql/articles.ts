import gql from 'graphql-tag';

const ArticleQuery = gql`

    query ArticleQuery($issue: ID) {
        issues(limit: 1, num: $issue) {
          max
          num
          articles {
            tags
            url
            id
            displayOrder
            dateCreated
            views
            author {
                fullName
                profileLink
            }
          }
        }
    }
`;

const ArticlePreviewIssueQuery = gql`

    query ArticleIssueQuery($issue: ID) {
        issues(limit: 1, num: $issue) {
          articles {
            url
            issue
            displayOrder
            views
            lede
            images(slide: true) {
                url
            }
          }
        }
    }
`;

const ArticlePreviewTagQuery = gql`

    query ArticleTagQuery($issue: Int, $tag: String) {
        articles(issue: $issue, tag: $tag) {
          url
          issue
          displayOrder
          views
          lede
          images(slide: true) {
              url
          }
        }
    }
`;

const ArticleUpdate = gql`
    mutation ArticleUpdate($data: [UpdateArticle], $password: String) {
        updateArticles(data: $data, password: $password) {
            id
            tags
            displayOrder
        }
    }
`;

const EditArticle = gql`
    mutation EditArticle($id: ID, $article: String) {
        editArticle(id: $id, article: $article) {
            id
            tags
            displayOrder
        }
    }
`;

const ArticleDelete = gql`

    mutation deleteArticles($ids: [ID], $password: String!) {
        deleteArticles(ids: $ids, password: $password) {
            id
        }
    }

`;

export {
    ArticleQuery,
    ArticlePreviewIssueQuery,
    ArticlePreviewTagQuery,
    ArticleUpdate,
    EditArticle,
    ArticleDelete
};