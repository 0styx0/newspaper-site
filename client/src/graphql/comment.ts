import { gql } from 'react-apollo';

const CommentCreate = gql`

    mutation CommentCreate($artId: ID!, $content: String!) {
        createComment(artId: $artId, content: $content) {
            id
        }
    }
`;

export {
    CommentCreate
};