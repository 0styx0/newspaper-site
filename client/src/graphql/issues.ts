import gql from 'graphql-tag';


const IssueQuery = gql`
    query IssueQuery {
        issues {
            num
            name
            views
            datePublished
            public
            canEdit
        }
    }
`;

const IssueUpdate = gql`
    mutation updateIssue($name: String, $public: Boolean, $password: String!) {
        updateIssue(name: $name, public: $public, password: $password) {
            name
            public
        }
    }
`;

const IssueInfoQuery = gql`
    query IssueInfo($num: ID) {
      issues(num: $num, limit: 1) {
          num
          max
          name
      }
    }
`;



export {
    IssueQuery,
    IssueInfoQuery,
    IssueUpdate
};
