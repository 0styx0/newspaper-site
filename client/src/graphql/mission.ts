import gql from 'graphql-tag';

const MissionEdit = gql`
    mutation editMission($mission: String!) {
        editMission(mission: $mission) {
            mission
        }
    }
`;

const MissionQuery = gql`
    query missionQuery {
        mission {
            mission
            canEdit
        }
    }
`;

export {
    MissionEdit,
    MissionQuery
};