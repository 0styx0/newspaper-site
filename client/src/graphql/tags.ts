import gql from 'graphql-tag';

const TagQuery = gql`

    query allTags {
        allTags
    }
`;

const TagCreate = gql`

    mutation createTag($tag: String!) {
        createTag(tag: $tag) {
            tag
        }
    }
`;

export {
    TagQuery,
    TagCreate
};
