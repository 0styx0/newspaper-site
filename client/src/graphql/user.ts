import { gql } from 'react-apollo';

const UserQuery = gql`

    query users($profileLink: String) {
        users(profileLink: $profileLink) {
            articles {
                id
                url
                dateCreated
                tags
                views
                issue
                canEdit
            }
            views
            level
            id
            fullName
            canEdit
        }
    }
`;

const PrivateUserQuery = gql`

    query users($profileLink: String) {
        users(profileLink: $profileLink) {
            notifications
            twoFactor
            email
            id
        }
    }
`;

const UserUpdate = gql`
    mutation updateProfile(
        $notifications: Boolean, $twoFactor: Boolean, $newPassword: String, $password: String!
    ) {
        updateProfile(
            notifications: $notifications, twoFactor: $twoFactor,
            newPassword: $newPassword, password: $password
        ) {
            id
        }
    }
`;

const RecoverPassword = gql`
    mutation recoverPassword($email: String!, $authCode: String!, $username: String!) {
        recoverPassword(email: $email, authCode: $authCode, username: $username) {
            message
        }
    }
`;

const UserLogin = gql`
    mutation login($username: String!, $password: String!) {
        login(username: $username, password: $password) {
            jwt
        }
    }
`;

// to be used only after user is logged in
const UserVerifyEmail = gql`
    mutation verifyEmail($authCode: String!) {
        verifyEmail(authCode: $authCode) {
            jwt
        }
    }
`;

const UserCreate = gql`
    mutation createUser(
        $username: String!,
        $email: String!,
        $level: Int,
        $firstName: String!,
        $middleName: String,
        $lastName: String!,
        $password: String!) {

            createUser(
                username: $username,
                email: $email,
                level: $level,
                firstName: $firstName,
                middleName: $middleName,
                lastName: $lastName,
                password: $password
            ) {
                id
            }
        }
`;

export {
    UserQuery,
    PrivateUserQuery,
    UserUpdate,
    UserLogin,
    RecoverPassword,
    UserCreate,
    UserVerifyEmail
};