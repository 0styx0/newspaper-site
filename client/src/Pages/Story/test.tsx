import casual from '../../tests/casual.data';
import { Story } from './shared.interfaces';
import { Comment } from '../../components/CommentList/shared.interface';
import StoryContainerWithGraphql, { StoryContainer } from './container';

import * as React from 'react';

import createHistory from 'history/createBrowserHistory';
import { renderWithGraphql } from '../../tests/snapshot.helper';
import mockGraphql, { createMutation, mountWithGraphql } from '../../tests/graphql.helper';
import { ArticleQuery } from '../../graphql/article';
import snapData from './__snapshots__/stories.example';
import setFakeJwt from '../../tests/jwt.helper';
import { setupComponent } from '../../tests/enzyme.helpers';
import { EditArticle } from '../../graphql/articles';
import { ReactWrapper } from 'enzyme';
import * as sinon from 'sinon';
const wait = require('waait');

setFakeJwt( { level: 3 } );
const history = createHistory();

type CustomCasual = typeof casual & {
    comments: Comment[],
    story: Story,
    data: {
        articles: Story
    }
};

const customCasual = casual as CustomCasual;

function setHistory() {

    const issue = casual.randomPositive;
    const url = casual.string; // not articleUrl since that's encoded
    history.push(`/stories/${issue}/url/${url}`);

    return {
        issue,
        url
    };
}

setHistory();

customCasual.define('story', (): Story & { displayOrder: number } => {

    return {
        id: casual.word,
        article: casual.article,
        canEdit: true,
        comments: customCasual.comments,
        tags: casual.tags,
        displayOrder: casual.randomPositive
    };
});

customCasual.define('data', () => {
    return {
        articles: [customCasual.story]
    };
});

const waitTime = 4;
describe('<StoryContainer>', () => {

    function setupQuery() {

        const data = customCasual.data;
        const { issue, url } = setHistory();

        return {
            query: createMutation(ArticleQuery, { issue, url }, data),
            data,
            issue,
            url
        };
    }

    async function setup(graphql: any[] = []
    ) {

        const wrapper = await mountWithGraphql(
            graphql,
            <StoryContainerWithGraphql />,
            waitTime
        );

        return wrapper;
    }

    function edit(
        wrapper: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>,
        selector: string,
        content: string
    ) {

        const elt = wrapper.find(selector);
        elt.simulate('blur', { target: { innerHTML: content } });

        const component = setupComponent(wrapper, StoryContainer);

        return component;
    }

    describe('snapshots', () => {

        async function testSnap(canEdit: boolean) {

            const { issue, url } = setHistory();
            const snapCopy = JSON.parse(JSON.stringify(snapData));
            snapCopy.canEdit = canEdit;

            await renderWithGraphql(
                mockGraphql(
                    [
                        createMutation(ArticleQuery, { issue, url }, { articles: [snapCopy] }),
                    ],
                    <StoryContainerWithGraphql />
                    ),
                waitTime
            );
        }

        it('does not allow edits when props.canEdit is false', async () => {

            await testSnap(false);
        });

        it('does allow edits when props.canEdit is true', async () => {

            await testSnap(true);
        });
    });

    describe('#onSaveEdits', () => {

        async function setupEditTests(selector: string, content: string, stateKey: string) {

            const wrapper = await setup([
                setupQuery().query
            ]);

            // const content = casual.articleHeader;
            const component = edit(wrapper, selector, content);
            expect(component.state[stateKey]).toBe(content);
        }

        it('saves header to state.header when header is edited', async () => {

            const content = casual.articleHeader;
            await setupEditTests('header', content, 'heading');
        });

        it('saves body to state.body when body is edited', async () => {

            const content = casual.articleBody;
            await setupEditTests('.storyContainer', content, 'body');
        });
    });

    describe('#onSubmit', () => {

        it('does not submit article if article has not changed', async () => {

            const wrapper = await setup([
                setupQuery().query
            ]);

            const component = setupComponent(wrapper, StoryContainer);
            component.onSubmit = () => expect(false).toBeTruthy();
            wrapper.find('.editable-submit').simulate('click');
        });

        it('submits the edited article when submit button is clicked', async () => {

            const newHeader = casual.articleHeader;
            const newBody = casual.articleBody;
            const { query, data } = setupQuery();
            const article = data.articles[0];

            const spy = sinon.spy(StoryContainer.prototype, `onSubmit`);

            const wrapper = await setup([
                query,
                createMutation(
                    EditArticle,
                    { id: article.id, article: newHeader + newBody },
                    {
                        editArticle: {
                            id: article.id, tags: article.tags, displayOrder: article.displayOrder
                        }
                    }
                )
            ]);

            edit(wrapper, 'header', newHeader);
            edit(wrapper, '.storyContainer', newBody);

            wrapper.find('.editable-submit').simulate('click');
            await wait(waitTime);

            expect(spy.calledOnce).toBeTruthy();
            (StoryContainer.prototype.onSubmit as any).restore();
        });
    });
});
