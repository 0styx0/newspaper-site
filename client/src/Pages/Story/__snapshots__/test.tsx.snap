// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StoryContainer> snapshots does allow edits when props.canEdit is true 1`] = `
<div>
  <div
    id="tags"
  >
    Tag(s): 
    hi, farewell
  </div>
  <article
    id="story"
  >
    <div>
      <div
        id="buttonContainer"
      >
        <button
          className="bold"
          onClick={[Function]}
          type="button"
        >
          bold
        </button>
        <button
          className="italic"
          onClick={[Function]}
          type="button"
        >
          italic
        </button>
        <button
          className="underline"
          onClick={[Function]}
          type="button"
        >
          underline
        </button>
        <button
          className="heading"
          onClick={[Function]}
          type="button"
        >
          heading
        </button>
        <button
          className="fontSize"
          onClick={[Function]}
          type="button"
        >
          fontSize
        </button>
        <button
          className="insertOrderedList"
          onClick={[Function]}
          type="button"
        >
          insertOrderedList
        </button>
        <button
          className="insertUnorderedList"
          onClick={[Function]}
          type="button"
        >
          insertUnorderedList
        </button>
        <button
          className="justifyFull"
          onClick={[Function]}
          type="button"
        >
          justifyFull
        </button>
        <button
          className="justifyLeft"
          onClick={[Function]}
          type="button"
        >
          justifyLeft
        </button>
        <button
          className="justifyRight"
          onClick={[Function]}
          type="button"
        >
          justifyRight
        </button>
        <button
          className="justifyCenter"
          onClick={[Function]}
          type="button"
        >
          justifyCenter
        </button>
        <button
          className="strikeThrough"
          onClick={[Function]}
          type="button"
        >
          strikeThrough
        </button>
        <button
          className="subscript"
          onClick={[Function]}
          type="button"
        >
          subscript
        </button>
        <button
          className="superscript"
          onClick={[Function]}
          type="button"
        >
          superscript
        </button>
        <button
          className="insertImage"
          onClick={[Function]}
          type="button"
        >
          insertImage
        </button>
        <button
          className="createLink"
          onClick={[Function]}
          type="button"
        >
          createLink
        </button>
        <button
          className="unlink"
          onClick={[Function]}
          type="button"
        >
          unlink
        </button>
        <button
          className="hideFromPreview"
          onClick={[Function]}
          type="button"
        >
          hideFromPreview
        </button>
        <br />
        <button
          className="editable-submit"
          onClick={[Function]}
          type="button"
        >
          Submit
        </button>
      </div>
      <header
        contentEditable={true}
        dangerouslySetInnerHTML={
          Object {
            "__html": "<h1>Title</h1><h4>Author</h4>",
          }
        }
        onBlur={[Function]}
      />
    </div>
    <div>
      <section
        className="storyContainer"
        contentEditable={true}
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>Content</p>",
          }
        }
        onBlur={[Function]}
      />
    </div>
  </article>
  <hr />
  <div
    id="comments"
  >
    <div
      id="comments"
    >
      <article
        className="comment"
      >
        <a
          className="author"
          href="/u/et"
          onClick={[Function]}
        >
          Cristian
        </a>
        <div
          className="content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "snarky comment",
            }
          }
        />
        <button
          className="deleteReply"
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </article>
      <article
        className="comment"
      >
        <a
          className="author"
          href="/u/Glib"
          onClick={[Function]}
        >
          Becky
        </a>
        <div
          className="content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "snarky comment 2",
            }
          }
        />
        <button
          className="deleteReply"
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </article>
      <article
        className="comment"
        id="reply"
      >
        <div
          className="content"
          onBlur={[Function]}
        />
      </article>
    </div>
  </div>
</div>
`;

exports[`<StoryContainer> snapshots does not allow edits when props.canEdit is false 1`] = `
<div>
  <div
    id="tags"
  >
    Tag(s): 
    hi, farewell
  </div>
  <article
    id="story"
  >
    <header
      dangerouslySetInnerHTML={
        Object {
          "__html": "<h1>Title</h1><h4>Author</h4>",
        }
      }
      onBlur={[Function]}
    />
    <section
      className="storyContainer"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Content</p>",
        }
      }
      onBlur={[Function]}
    />
  </article>
  <hr />
  <div
    id="comments"
  >
    <div
      id="comments"
    >
      <article
        className="comment"
      >
        <a
          className="author"
          href="/u/et"
          onClick={[Function]}
        >
          Cristian
        </a>
        <div
          className="content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "snarky comment",
            }
          }
        />
        <button
          className="deleteReply"
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </article>
      <article
        className="comment"
      >
        <a
          className="author"
          href="/u/Glib"
          onClick={[Function]}
        >
          Becky
        </a>
        <div
          className="content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "snarky comment 2",
            }
          }
        />
        <button
          className="deleteReply"
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </article>
      <article
        className="comment"
        id="reply"
      >
        <div
          className="content"
          onBlur={[Function]}
        />
      </article>
    </div>
  </div>
</div>
`;
